teps/add-variable-retries-and-retry-count.md

```diff
@@ -0,0 +1,105 @@
+---
+title: add-variable-retries-and-retry-count
+authors:
+  - "@zhangyumeng"
+creation-date: 2020-10-14
+last-updated: 2020-10-19
+status: proposed
+---
+
+# TEP-0026: Add Variable `retries` and `retry-count`
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Alternatives](#Alternatives)
+- [References](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes adding `retries` and current retry times to Pipeline variables.
+
+## Motivation
+
+PipelineTask can be rerun after failure when user specifies `retries` in PipelineTask. In that case, exposing the retry count can give user more information. User can dispatch a notification after certain times of failure, or customize logs according to retry count.
+
+### Goals
+
+- add 2 variables indicating retries and retry count of a PipelineTask
+
+### Non-Goals
+
+- execute different PipelineTask based on retry count
+- emit an event on every retry
+
+## Proposal
+
+There are 2 variable substitutions `tasks.<taskName>.retries` and `tasks.<taskName>.retry-count` needed to be supported:
+
+1. For retrieving `retries`, we can use `Retries` that stored in `PipelineTask` struct
+2. For retrieving `retry-count`, we can use length of `RetriesStatus` in `TaskRunStatus`
+
+Code example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: failed-task
+spec:
+  params:
+    - name: num-retries
+      type: string
+    - name: max-retries
+      type: string
+  steps:
+    - image: alpine:3.12.0
+      script: |
+        #!/usr/bin/env sh
+        if [ $(params.num-retries) == $(params.max-retries) ]; then
+          echo "This is the last retry. \n"
+        fi
+    - image: ubuntu
+      script: |
+        #!/usr/bin/env sh
+        /path/to/<program-that-fails>
```

Comment:
it's interesting that this example has to be custom built to take these values as a param - if you wanted to meet the use case you mentioned and alert slack in the case of failure, you'd need one Task that both a) does the thing that needs to be retried and b) alerts slack - unless we let Tasks access the retry count of other running Tasks @_@

Reply:
Sorry for the late response! I agree with your suggestion of limiting `retries` to the task own context. I should have added this implementation to the doc and made the doc more detailed. Did we decide to not move forward with this tep since users can use pipelines + trigger to send the notification for now and pipelines should have a more decent way to notify the users([tektoncd/pipeline#1740](https://github.com/tektoncd/pipeline/issues/1740)). I'm glad to help if there's anything I can do to make pipelines support notifications.

Reply:
> I'm glad to help if there's anything I can do to make pipelines support notifications.

@afrittoli has now opened https://github.com/tektoncd/community/pull/275 about proposing a notifications feature :D

> Did we decide to not move forward with this tep since users can use pipelines + trigger to send the notification for now and pipelines should have a more decent way to notify the users(tektoncd/pipeline#1740)

I don't think we decided definitively - it'd be interesting to hear from @IdanAdar who opened https://github.com/tektoncd/pipeline/issues/2725 to hear his thoughts. i think there's a good chance we could add the simplified version of this variable interpolation feature in the meantime. @yaoxiaoqi is this something you still want to work on or do you want to close this PR for now?

Reply:
I'd like to continue to work on this issue if we decide this can be implemented. If this decision can't be made shortly this PR can also be closed.